### DISCUSSION TITLE
How to achieve group/global opacity in 2d?

### DISCUSSION THREAD

**Discussion by zimond** [bevyengine/bevy#13831] [Q&A]
Let's say I have two sprites in a group. I want the parent entity to be of `0.5` alpha, currently in bevy it seems that all I can do is to set sprite color to 0.5 alpha on each sprite. However, they will blend since they are both transparent now, which is not what I want (opaque blend then 0.5 alpha as a whole piece).

what I get:
<img width="258" alt="Screenshot 2024-06-14 at 12 02 31 AM" src="https://github.com/bevyengine/bevy/assets/4921289/521abdec-6d85-441d-bb63-3a7ef1792ada">

what I want:
<img width="258" alt="Screenshot 2024-06-14 at 12 03 12 AM" src="https://github.com/bevyengine/bevy/assets/4921289/7a1ef711-4fab-47de-a9d1-d70efe0efb1f">

**Comment by bugsweeper** [comment]
I think you can't achieve this by just alpha channel in color, because at the moment, when color applies alpha it takes just a color under it, not the sprite or background, if input colors for two pixels aren't the same (first figure is not fully transparent) the resulting color won't be the same. If background has same color you can make sprites with desired color with no alpha, then overlapping region would have same color.
For understanding the problem try to answer the question: what color of overlapping region should be if ellipse has red base color and triangle has blue base color?
OR if you want to use primitive shapes, than you can try to generate meshes from shapes, than you can make single mesh from two by [merging](https://dev-docs.bevyengine.org/bevy/render/prelude/struct.Mesh.html#method.merge) it. I think if it is single mesh you can achieve desirable effect.

**Reply by zimond** [reply]
> what color of overlapping region should be if ellipse has red base color and triangle has blue base color?

It should be blue. Do you think this is doable via a custom fragment shader or something?

**Reply by bugsweeper** [reply]
> It should be blue.

So desired logic is mixing colors only with background, but not surfaces under the sprite. Of course it's doable! You can do it with your fragment shader, 1. you should just make custom material, 2. pass background color (or hardcode it in shader) and alpha channel or full rgba color (or hardcode it in shader). 3. In the shader you would just mix that colors. Engine will see 100% opaque sprite, user will think that sprite is 50% transparent. If backgound is not a single color but texture, then 2. the material should keep background texture and position of the sprite to calculate which background pixel should be mixed.

**Reply by zimond** [reply]
what if the two sprites and many more sprites beneath them are all moving? How can I correctly pass background color to shader? Could #2223 be related to this?
